### PR TITLE
Explicitly set permission on the socket file created by DiagnosticServer

### PR DESCRIPTION
**Pull request by sywhang**
This changes the permission set on the diagnostics IPC socket file from 775 to 700.

### CODE REVIEW

**Comment by josalem** [comment]
CC @shirhatti

File: src/debug/debug-pal/unix/diagnosticsipc.cpp [hunk 1]
```diff
@@ -60,6 +61,11 @@ IpcStream::DiagnosticsIpc *IpcStream::DiagnosticsIpc::Create(const char *const p
         pd.m_ApplicationGroupId,
         "socket");
 
+    if (fchmod(serverSocket, S_IRWXU) == -1)
```
Comment (jorive) [inline src/debug/debug-pal/unix/diagnosticsipc.cpp]: Don't we want to have assert here too?
```cpp
if (callback != nullptr)
    callback(strerror(errno), errno);
_ASSERTE(!"Failed to set permissions");
```

Comment (jorive) [reply src/debug/debug-pal/unix/diagnosticsipc.cpp]: Maybe no exec permission?

**Comment by jorive** [comment]
On OSX x64_checked:
```log
ASSERT FAILED
	Expression: !"Failed to set permissions on diagnostics IPC socket."
	Location:   line 66 in /Users/vsts/agent/2.150.3/work/1/s/src/debug/debug-pal/unix/diagnosticsipc.cpp
	Function:   Create
	Process:    55131
```

File: src/debug/debug-pal/unix/diagnosticsipc.cpp [hunk 2]
```diff
@@ -60,6 +61,12 @@ IpcStream::DiagnosticsIpc *IpcStream::DiagnosticsIpc::Create(const char *const p
         pd.m_ApplicationGroupId,
         "socket");
 
+    if (fchmod(serverSocket, S_IRUSR | S_IWUSR) == -1)
+    {
+        _ASSERTE(!"Failed to set permissions on diagnostics IPC socket.");
```
Comment (jorive) [inline src/debug/debug-pal/unix/diagnosticsipc.cpp]: Maybe we could with `strerror(errno)` on the string message?

Comment (jorive) [reply src/debug/debug-pal/unix/diagnosticsipc.cpp]: I would also add the callback call (we are logging to STRESS_LOG)

**Comment by jorive** [comment]
From [fchmod.2](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/fchmod.2.html)
```
     fchmod() will fail if:

     [EBADF]            fildes is not a valid file descriptor.

     [EINVAL]           fildes refers to a socket, not to a file.

     [EINVAL]           mode is not a valid file mode.

     [EINTR]            Its execution was interrupted by a signal.

     [EIO]              An I/O error occurred while reading from or writing to
                        the file system.

     [EPERM]            The effective user ID does not match the owner of the
                        file and the effective user ID is not the super-user.

     [EROFS]            The file resides on a read-only file system.
```

**Comment by jorive** [comment]
Optionally, @josalem suggested placing the sockets under a directory with the right permissions.
Maybe something like:
```
/tmp/dotnet-diagnostics/<PID>.socket
```

**Comment by sywhang** [comment]
I've been debugging this all morning - It seems like in OSX you can't set permissions on a socket. I have a workaround for it that I'm testing now. I'd prefer to not create a directory - In that case we have to carry over a debt from P5/P6 on CLI tools and if we can avoid it, we should.

**Comment by jorive** [comment]
@sywhang Don't you have to call `umask(prev_mask);` in all potential return calls of the method?

**Pull request merged**